src/storage/sqlite.js

```diff
@@ -88,6 +94,23 @@ export class SQLDatabase {
   }
 }
 
+// An absurd little workaround for expo-sqlite, or really
+// the @expo/websql library under it, being too eager to check a
+// transaction's queue and declare it complete.
+async function keepQueueLiveWhile(
+  tx: WebSQLTransaction,
+  f: () => void | Promise<void>,
+): Promise<void> {
+  let done = false;
+  const hold = () => tx.executeSql('SELECT 1', [], () => (done ? undefined : hold()));
```

Comment:
I guess it'd be pretty pathological for a future/different `@expo/websql` [implementation](https://github.com/expo/node-websql/blob/fa169c167/lib/websql/WebSQLTransaction.js#L95-L105) to have `executeSql` call the callback synchronously. When that occurred to me, I was more convinced that we can't have an infinite recursion problem here. (Right?)

I wonder about a strain on performance caused by these rapid-fire 'SELECT 1' statements at all the different layers. There will be more two-way traffic on the React Native bridge; I'm not sure what that might mean for memory. Worth at least manually testing how it feels when we're writing something huge, like users on CZO?

Reply:
> I guess it'd be pretty pathological for a future/different `@expo/websql` [implementation](https://github.com/expo/node-websql/blob/fa169c167/lib/websql/WebSQLTransaction.js#L95-L105) to have `executeSql` call the callback synchronously. When that occurred to me, I was more convinced that we can't have an infinite recursion problem here. (Right?)

Yeah, that would be quite pathological. Generally in a callback-based API, it's a principle of good API behavior that if the callback's invocation is usually asynchronous, or ever asynchronous, or the API looks like it might be asynchronous, then it's only ever invoked asynchronously and never synchronously, i.e. it's never invoked before the API function has returned to its caller. When a callback is sometimes unexpectedly called synchronously, it's tricky for users of the API to avoid having subtle bugs.

There's a bit of discussion of this on a page of Node's docs that I ran across while trying to understand `process.nextTick`:
https://nodejs.org/en/docs/guides/event-loop-timers-and-nexttick/#why-would-that-be-allowed

More concretely, the Web SQL Database spec which this library generally aims to follow specifically requires that the statement just gets queued up, not synchronously executed:
https://www.w3.org/TR/webdatabase/#executing-sql-statements
and so a fortiori the statement callbacks.

Reply:
> I wonder about a strain on performance caused by these rapid-fire 'SELECT 1' statements at all the different layers. There will be more two-way traffic on the React Native bridge; I'm not sure what that might mean for memory. Worth at least manually testing how it feels when we're writing something huge, like users on CZO?

Indeed.

Note that as one mitigating factor, the statements are run in order -- so when we queue one of these `SELECT 1` statements after some other statement, its callback won't get run (and possibly put another one on the queue) until that other statement completes. I believe that means that in a transaction with N actual app-level statements, this workaround will only add N+2 of these queue-keepalive `SELECT 1` statements, no matter how big or slow the app-level statements are.

Reply:
> when we queue one of these SELECT 1 statements after some other statement, its callback won't get run (and possibly put another one on the queue) until that other statement completes

Mm, interesting, I didn't think of that. That's good, I guess!

> I believe that means that in a transaction with N actual app-level statements, this workaround will only add N+2 of these queue-keepalive `SELECT 1` statements, no matter how big or slow the app-level statements are.


I think I'm counting N+1. Where do you get the 2 in N+2?

Reply:
> I think I'm counting N+1. Where do you get the 2 in N+2?

Hmm, not sure. When I think through it again now I only see N+1 :slightly_smiling_face: -- one before each app-level statement, and one at the end.

Also, that is a valid upper bound but here's a tighter one: if the app-level code `await`s a `tx.executeSql` only M times, then we'll have at most M+1 of these `SELECT 1` statements. In particular, if you have code like
```js
  db.transaction(tx => {
    tx.executeSql(…);
    tx.executeSql(…);
    tx.executeSql(…);
    tx.executeSql(…);
  });
```
where you just issue a bunch of statements but don't wait for any of their results, then we'll only do `SELECT 1` once.

> Also, if one of these `SELECT 1` statements is alone in the queue and other app-level statements aren't being enqueued, can't we accidentally break such a bound by running `SELECT 1` over and over until another app-level statement gets enqueued and gets its turn? I think that could happen if there's some async non-statement work we're waiting for, like compression/decompression steps you mention in the issue description, for migrations.

Ah, yeah, my comment above was stated too broadly. That's right -- this bound is really about what happens if the app-level code is doing a bunch of SQL statements but not doing anything *else* that waits for I/O (or for a timer); in other words, if the only things it ever `await`s are either a `tx.executeSql(…)` or a promise that's already resolved.


> I'm not sure what if any noticeable strain on performance these `SELECT 1` statements will have. Possibly none or not much.

I think a key thing in our usage will be that outside of migrations, all our transactions will be entirely synchronous: they're all inside the replacement `AsyncStorage`, and those never have a transaction callback wait for anything before enqueueing all the statements it wants to do. So they only ever have a single `SELECT 1` added here.

That means the performance impact should be basically limited to migrations. It's pretty OK if those run somewhat slower, because they're infrequent. A migration would have to take over a second for its performance to start really being a problem. Compressing or decompressing a few megabytes should take a very small fraction of that, so it'd have to be a very large slowdown to be a problem.

Reply:
Cool, sounds good! Thanks!